### PR TITLE
Sign up integration

### PR DESCRIPTION
**Pull request by ricardoamoreno**
## Summary

- /users endpoint implemented
- ajv implemented to validate email, password, firstName and lastName fields
- bcrypt implemented to encrypt user password
- persisted user data on postgres database

## Known Issues

none

## Trello Card

[Sign Up](https://trello.com/c/X5k0cjom)

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+const { hashPassword } = require('../helpers/encrypt');
+const { createUser } = require('../services/users');
+
+exports.signUp = (req, res, next) => {
+  const { firstName, lastName, email, password } = req.body;
+  return hashPassword(password)
+    .then(hashPass => createUser({ firstName, lastName, email, password: hashPass }))
+    .then(user => {
+      logger.info(`user ${user.firstName} was created succesfully`);
```
Comment (martinremis) [inline app/controllers/users.js]: Here it would be better to log the user id instead of the name

File: app/middlewares/errors.js [hunk 2]
```diff
@@ -7,7 +7,9 @@ const statusCodes = {
   [errors.DATABASE_ERROR]: 503,
   [errors.DEFAULT_ERROR]: 500,
   [errors.EXTERNAL_API_ERROR]: 503,
-  [errors.NOT_FOUND_ERROR]: 404
+  [errors.NOT_FOUND_ERROR]: 404,
+  [errors.VALIDATION_ERROR]: 422,
```
Comment (martinremis) [inline app/middlewares/errors.js]: For a schema validation error, is better to use error code 400. The code 422 is more for a correct schema but some  further validation error.

File: app/middlewares/validate-dto.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const { validationError } = require('../errors');
+
+function validateDto(ajvValidate) {
```
Comment (martinremis) [inline app/middlewares/validate-dto.js]: Why is this defined as a classic function? Why not a function in a variable? Why not an anonymous function (arrow function)?

Comment (martinremis) [reply app/middlewares/validate-dto.js]: I am not saying it is bad, just want to talk about the differences.

File: app/models/user.js [hunk 4]
```diff
@@ -1,15 +1,19 @@
...
       email: { type: DataTypes.STRING, allowNull: false, field: 'email', unique: true },
       password: { type: DataTypes.STRING, allowNull: false }
     },
     {
-      timestamps: false
+      timestamps: true,
+      underscored: true
     }
   );
+  User.createUser = user => User.create({ ...user });
```
Comment (martinremis) [inline app/models/user.js]: We usually do not save this methods here but in a service. It is not bad to do it like this, but for convention we use to have a "logicless" model.

File: app/services/users.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const { User } = require('../models');
+
+const { databaseError } = require('../errors');
+
+exports.createUser = user =>
+  User.createUser(user).catch(error => {
```
Comment (martinremis) [inline app/services/users.js]: You can add some logs here

**Review by martinremis — CHANGES_REQUESTED** [review]

File: app/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+const logger = require('../logger');
+
+const { hashPassword } = require('../helpers/encrypt');
+const { createUser } = require('../services/users');
+
+exports.signUp = (req, res, next) => {
```
Comment (wgermancutraro) [inline app/controllers/users.js]: This is good, but async / await better :)

File: app/controllers/users.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  const { firstName, lastName, email, password } = req.body;
+  return hashPassword(password)
```
Comment (wgermancutraro) [inline app/controllers/users.js]: Same, async / await

File: app/helpers/encrypt.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+const bcrypt = require('bcryptjs');
+const logger = require('../logger');
+const { encryptionError } = require('../errors');
+
+exports.hashPassword = password =>
```
Comment (wgermancutraro) [inline app/helpers/encrypt.js:5]: Same, async / await

File: app/middlewares/errors.js [hunk 9]
```diff
@@ -7,7 +7,9 @@ const statusCodes = {
...
+  [errors.ENCRYPTION_ERROR]: 500
```
Comment (wgermancutraro) [inline app/middlewares/errors.js]: I would use an "internal error" constant name.

File: app/middlewares/validate-dto.js (hunk 3 above)
Comment (ricardoamoreno) [reply app/middlewares/validate-dto.js]: mmm, I didn't think about it, I know that the difference between classic functions and arrow functions is the scope of "this" and that arrow functions have an implicit return. I am not sure but, I think that for this case it doesn't affect. I rewrote the function in this way using arrow functions, do you think it is better?    

```
  const validateDto = ajvValidate => (req, res, next) => {
  const valid = ajvValidate(req.body);
  if (!valid) {
    const { errors } = ajvValidate;
    return next(validationError(errors));
  }
  return next();
};
```

File: app/models/user.js (hunk 4 above)
Comment (ricardoamoreno) [reply app/models/user.js]: Understood

File: app/controllers/users.js (hunk 6 above)
Comment (ricardoamoreno) [reply app/controllers/users.js]: I agree, it was already changed

File: app/middlewares/errors.js (hunk 9 above)
Comment (ricardoamoreno) [reply app/middlewares/errors.js]: I added it

File: app/controllers/users.js (hunk 1 above)
Comment (ricardoamoreno) [reply app/controllers/users.js]: I changed it

File: app/middlewares/validate-dto.js (hunk 3 above)
Comment (martinremis) [reply app/middlewares/validate-dto.js]: I think it is the same for this case, but we usually use const vars for functions declarations, even for anon or classic type. It was just a question to see if you know that 😛

**Review by martinremis — DISMISSED** [review]
You have conflicts with the base branch. You may need to update this one up to weets-api-integration

**Comment by martinremis** [comment]
And remember **not to merge** until changing base to main again

File: package.json [hunk 10]
```diff
@@ -66,7 +66,10 @@
     }
   },
   "dependencies": {
+    "ajv": "^8.6.3",
+    "ajv-formats": "^2.1.1",
     "axios": "^0.19.2",
+    "bcrypt": "^5.0.1",
```
Comment (martinremis) [inline package.json]: Are you using this new library?

**Review by martinremis — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (ricardoamoreno) [reply package.json]: No, sorry, I forgot delete it

**Review by martinremis — APPROVED** [review]

**Pull request merged**